Added V2 and ISA support

tnemoz opened the pull request:
### Summary
This PR's goal is to adapt the whole code base to support V2 primitives and ISA circuits. Will fix #136, fix #164, fix #165, fix #194, fix #204, fix #213, fix #220, fix #224, and fix #229.

### What's working

| | Qiskit 1.4/2.0 | Qiskit 2.2 | Comments |
| ----- | --------- | -------| ----- |
| `eigensolvers` | :heavy_check_mark: | :heavy_check_mark: | |
| `gradients` | :heavy_check_mark: | :heavy_check_mark:  | A test is skipped as it takes quite some time, and we're not sure to see its point at the moment. The whole test suite takes about 20 minutes to run. |
| `minimum_eigensolvers` | :heavy_check_mark: | :heavy_check_mark: | |
| `optimizers` | :heavy_check_mark: | :heavy_check_mark:  |  |
| `state_fidelities` | :heavy_check_mark: | :heavy_check_mark: |  |
| `time_evolvers` | :heavy_check_mark: | | The tests for Qiskit 2.2 don't pass because https://github.com/Qiskit/qiskit/issues/14614 wasn't fixed, so we need to circumvent the problem |
| `utils` | :heavy_check_mark: | :heavy_check_mark: |  |
| `amplitude_estimators` | :heavy_check_mark: | :heavy_check_mark: | A test is skipped for now until Qiskit/qiskit#14250 is fixed and merged |
| `grover` | :heavy_check_mark: | :heavy_check_mark: | A test is skipped for now but should probably be removed or refactored, see https://github.com/qiskit-community/qiskit-algorithms/issues/136#issuecomment-2291169158. |
| `phase_estimators` | :heavy_check_mark: | :heavy_check_mark: |  |
| `validation` | :heavy_check_mark: | :heavy_check_mark: |  |

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=197) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=197) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/qiskit-community/qiskit-algorithms?pullRequest=197) it.</sub>

woodsp-ibm commented:
FYI. I merged a PR that fixed CI with the main branch and changes that will going into the shortly to be released 1.2.0 so that CI now passes fully here and does not fail with the two jobs that check things out against qiskit main, and updated the branch here so this now completely passes.

I think this is fine. I must admit though I had wondered, given that I think we only ever need something that has a run() method, that takes circuits and gives transpiled circuits back whether, something could be done around a type that was more targeted along those lines. The pass manager has a number of methods that I do not think we would call, and the [AI Transpiler Service](https://docs.quantum.ibm.com/guides/qiskit-transpiler-service) which has a run method could be used too but is not a passmanager - i,e, so that a Passmanager instance or the AI transpiler could be used - or anything else that had a run method that did the transpilation. I know Optimization went for BassPassManager too as the type and maybe thats ok.

Whatever is done I think we might want to note in the docstring for this that's its optional and that passing in None results in the circuit not being transpiled. Maybe for some algos like VQE one can already pass a circuit (ansatz) that is transpiled for a given backend so the note there might be a little different to accommodate that  - for PE they build their own circuits of course so that aspect does not apply.

woodsp-ibm commented on qiskit_algorithms/custom_types.py:
```diff
@@ -12,11 +12,11 @@
 from qiskit import QuantumCircuit
 
-_Circuits = Union[list[QuantumCircuit], QuantumCircuit]
+_Circuits = Union[List[QuantumCircuit], QuantumCircuit]
```
You can use `list` in 3.8 - which arguably is better and what we have tried to do elsewhere. You do need to add a future import to have that work `from __future__ import annotations`. You can see examples in this repo e.g. in this file which does that and uses it https://github.com/qiskit-community/qiskit-algorithms/blob/main/qiskit_algorithms/time_evolvers/pvqd/pvqd_result.py

tnemoz replied:
I initially went for 
```python
_Circuits = list[QuantumCircuit] | QuantumCircuit
```
but then `make lint` complained with:
```
************* Module qiskit_algorithms.custom_types
qiskit_algorithms/custom_types.py:21:12: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
```
which is why I went for `Union` instead of `|`. Or is there a workaround I don't know of?

woodsp-ibm replied:
That future import is needed for `|` as well - but if you had that import then as I mentioned I was not not sure here as I recall having issues when defining types before using these newer aspects where we had to have them defined with the former Typing constructs. If you search in this repo you will find just a few Union hits where they are pretty much limited to being used in type defines like this - I also see List being used in the ones here so it may well be the same issue with that. I commented mostly as a saw the commit that changed things to fix it for lint in 3.8 from list to List that was all the change I did not see a from future import being removed which is needed to type things this newer way in 3.8.

woodsp-ibm replied:
One further comment around 3.8. Qiskit has already deprecated support for Python 3.8 and support will be removed in 1.3.0 which is due in Nov. Python 3.8 EOL is Oct this year so among the changes here it could be a choice to drop 3.8 support too along with these changes. To drop this would evidently also include dropping it from CI tests etc and bumping things to run at 3.9 min. 

In talking about versions I will note 3.13 is planned to be available beginning of Oct. There is an issue on Qiskit Qiskit/qiskit#12903 to support this so again depending on timeline.... but hopefully that's just adding it as supported and to CI to test when dependencies are there and it just works!

tnemoz commented:
Hey @woodsp-ibm! Sorry for the time without any news, we've had quite a lot on our plates recently :(

We just wanted to keep you posted on what's left to do. Normally, all the checkboxes that are... well, checked in the first post of this PR are ready for review. Either their tests pass, or they will once the non-checked items are done. Doing these non-checked items shouldn't be too long, it's just a matter of finding the time to do so (but hopefully soon!). Once again, sorry for that.

The reason why we're pinging you even though it's not entirely finished is so that you can take a quick look at how things have been done. Since the code style is (normally) consistent between files, an error or bad coding pattern that you spot on a file will likely be repeated on others, so it's probably fine for now if you just look at "representative files" and leave comments on them, we'll just correct the other spots where we did the same mistake.

Sorry again for the long time without news, but we should be close to the end!

OkuyanBoga commented:
Hi all, 

I suggest looking at our version (not fully optimised yet) [gradients](https://github.com/qiskit-community/qiskit-machine-learning/tree/main/qiskit_machine_learning/gradients) and [state_fidelities](https://github.com/qiskit-community/qiskit-machine-learning/tree/main/qiskit_machine_learning/state_fidelities).

We will update these for Qiskit machine learning, which might help you to update the rest.

panaschristou commented:
Hello, I hope everybody is doing well. I wanted to know what is the current progress with this and if there is any way I can help even though I am not very experienced in open source. I really need this to work with primitives V2.

tnemoz commented:
Hey @panaschristou! Yes, it's still being worked on, sorry for the delays :(
The only thing missing for now are the minimum eigensolvers (i.e. VQE-like algorithms). All the others functionalities have been changed along with their tests, and the tests pass. In particular, if you need something else than minimum eigensolvers, then you can fork and use the [dev branch](https://github.com/LenaPer/qiskit-algorithms/tree/dev) of the repository.
We had quite long pauses and I apologize for that, but we're actively working on the minimum eigensolvers [on this branch](https://github.com/LenaPer/qiskit-algorithms/tree/minimum_eigensolvers), and hopefully we should be done by a week or so. One that done, the only things left to do would be to add the missing copyrights, formatting the code and of course correcting the feedback from the reviewers, but in the mean time you should still be able to use it one the minimum eigensolvers are done.
Once again I apologize for this :(

panaschristou commented:
No worries, thank you for your hard work and the quick reply, we all appreciate it!
I am actually using the VQE-like aglorithms so I can wait until you finish with those since it will only take a week or so like you said. Please reply here if possible when you push the changes for those and I will fork the dev branch.
Again, thank you for doing this!

woodsp-ibm commented:
Just noting, as much for me as anything, this that the `requirements.txt` should be updated as currently qiskit is pinned to < 2.0 so CI here passes as the main code still uses the V1 primitives,. Also the min requirement should be bumped as what is there currently predates the inclusion of V2 primitives.

DYMS-212 commented:
Hi everyone! I hope you're doing well.

I'm currently working on a project that requires using VQE and I'm very interested in using the primitives V2 implementation. I saw that the work on minimum eigensolvers is still ongoing, and I was wondering if there’s anything I can help with — even small tasks or testing — to speed things up. I’m also hoping to better understand how to work with the dev branch in the meantime.

Thanks in advance, and I really appreciate all the hard work you're putting into this!

tnemoz commented:
@panaschristou and @DYMS-212 
The minimum eigensolvers are done, I just pushed the latest changes to the [dev branch](https://github.com/LenaPer/qiskit-algorithms/tree/dev). All the tests are passing, and I've also updated the associated tutorials if you want to have a look :slightly_smiling_face:

woodsp-ibm commented:
I fixed the qnspa merge conflict here from where it had been updated here by #218. So I changed similarly to make sampler required again (not allow None) and then the new transpiler options are after the * now as optionals.

woodsp-ibm commented:
For _circuit_key. Maybe a simple expedient for now is to copy the code over that was there for _circuit_key and make a local function here https://github.com/Qiskit/qiskit/blob/stable/1.4/qiskit/primitives/utils.py called circuit_key() and use that 

As far as I can see that function uses public attributes of QuantumCircuit all except this line

> None if circuit._op_start_times is None else tuple(circuit._op_start_times)

but from QuantumCircuit code I see this, which says there is a public attribute added so no need to access the private member.

>    :attr:`op_start_times`         Start times of scheduled operations, added by scheduling
                                   transpiler passes.

FYI here is some link to an original discussion around circuit_key and it references the issue it was addressing https://github.com/Qiskit/qiskit/pull/8604

I will note Qiskit Machine Learning has copied some code over from Algs here that it was dependent upon, including code depending on that circuit key, so they will end up having the same issue. I'll reference the issue for their 2.0 migration which is https://github.com/qiskit-community/qiskit-machine-learning/issues/897

As to what to perhaps do instead, if this expedient would be done but wanted to be removed going forwards, maybe its worthy of opening a separate PR to propose solution/discuss?

woodsp-ibm commented:
FYI: In case it helps and you were not aware... in the root of the project is a make file with several targets,`'make copyright` will check and fix copyrights, `make lint`, `make mypy` run those over the code, `make spell` does what it implies as does `make html` if you want to see the docs that are created (it says where it puts the html). `make style` just does a style check whereas `make black` fixes the style. `make test` runs all the unit tests though often I use the IDE I'm using to do this.

tnemoz commented:
> FYI: In case it helps and you were not aware... in the root of the project is a make file with several targets,`make copyright` will check and fix copyrights, `make lint`, `make mypy` run those over the code, `make spell` does what it implies as does `make html` if you want to see the docs that are created (it says where it puts the html). `make style` just does a style check whereas `make black` fixes the style. `make test` runs all the unit tests though often I use the IDE I'm using to do this.

I am aware! Actually, this is what I've used to correct the linting errors in local. For the subsequent runs, I just always forget to rerun them :sweat_smile:

Concerning the test suite however, I do run it in local beforehand, but it's at the exception of the `gradients` tests, which seem to run *much* faster here than on my computer.

Is there a problem I'm not aware of with this approach (like if there's a monthly rate of workflows GitHub can run on a given repository, or if you receive a notification for each failed run)?

woodsp-ibm commented on qiskit_algorithms/amplitude_estimators/iae.py:
```diff
@@ -54,7 +55,9 @@ def __init__(
-        sampler: BaseSampler | None = None,
+        sampler: BaseSamplerV2 | None = None,
+        transpiler: Transpiler | None = None,
+        transpiler_options: dict[str, Any] | None = None,
```
If this is left as the default None (or explicitly passed as None of course) does it work - my impression is that it would be doing a **None when calling the transpiler method passed. The tests I saw did not cover this aspect - unless I missed things.

woodsp-ibm replied:
I am not quite sure why you resolved this. Does it not need the transpiler_options internally setting to {} like I see you did elsewhere and commented below.

woodsp-ibm commented on qiskit_algorithms/amplitude_estimators/iae.py:
```diff
@@ -271,8 +283,10 @@ def estimate(
+            warnings.warn(
+                "No sampler provided, defaulting to StatevectorSampler from qiskit.primitives"
+            )
+            self._sampler = StatevectorSampler()
```
This is just a comment as I can see you just altered what was there to V2. I have no idea why we allow None in constructor, which then is valid input as such to me, and then warn about it here what we do with None. Why we didn't just set this as the self._sampler in the constructor if sampler was passed was None and note this behavior in the docstring I do not know. That way the getter would always return a valid instance. And if we did the same in setter - ie allow None I guess the constructor could use that rather than duplicate handling None so internally its always a valid instance rather waiting until here in estimate to set it whereby the getter now starts returning this whereas it would have been None before. I guess an issue could be raised to "improve" this behavior given the goal was to go over to V2's which it has done here,

woodsp-ibm commented on qiskit_algorithms/gradients/base/base_estimator_gradient.py:
```diff
@@ -107,10 +101,12 @@ def run(
-                Higher priority setting overrides lower priority setting
+            precision: Precision to be used by the underlying Estimator. If a single float is
+                provided, this number will be used for all circuits. If a sequence of floats is
+                provided, they will be used on a per-circuit basis. If none is provided, the
```
> If none is provided, the
> gradient's default precision will be used for all circuits. If this number is
> also set to None, the underlying primitive's default precision will be used
> for all circuits.

I am trying to parse these - why are there two sentences and `none` in the first case I guess that means `None` or is that somehow an empty sequence? The "also" in the second sentence I guess I parse "as well as" which makes things hard for me to read/comprehend. Either which way the first sentence, if it has meaning, and None as per the second just mean the `the underlying primitive's default precision will be used for all circuits.` either way.

tnemoz replied:
I meant `none` in the sens of `None` indeed, it's missing the capitalization. Just to be sure I understand, does the following work for you?
`precision: Precision to be used by the underlying Estimator. If a single float is
                provided, this number will be used for all circuits. If a sequence of floats is
                provided, they will be used on a per-circuit basis. If not set, the
                gradient's default precision will be used for all circuits. If this number is
                also not set the underlying primitive's default precision will be used
                for all circuits.`

woodsp-ibm replied:
In the last part `If this number is also not set...` I guess the `this` refers to the `gradients default precision` number in the prior sentence. So if I do it like below is that the intent (where I guess if I pass a specific precision to the estimator that is used by default - hence I put brackets around the default since it will only use its own internal default, I think, in the case nothing is provided).

`If not set, the gradient's default precision will be used for all circuits., and if that is None (not set) then the underlying primitive's (default) precision will be used for all circuits.`

woodsp-ibm commented on qiskit_algorithms/gradients/base/base_sampler_gradient.py:
```diff
@@ -36,32 +33,31 @@
-                default options > primitive's default setting.
-                Higher priority setting overrides lower priority setting
+            shots: Number of shots to be used by the underlying Sampler. If provided, this number
+                takes precedence over the default precision of the primitive. If None, the default
```
I thought Sampler just allowed shots to be passed (None being some default). So I am trying to figure what this is with precision `If provided, this number takes precedence over the default precision of the primitive.`

tnemoz replied:
Bad copy/paste, sorry for that!

woodsp-ibm commented on qiskit_algorithms/gradients/base/base_sampler_gradient.py:
```diff
@@ -73,10 +69,12 @@ def run(
-                Higher priority setting overrides lower priority setting
+            shots: Number of shots to be used by the underlying sampler. If a single integer is
+                provided, this number will be used for all circuits. If a sequence of integers is
+                provided, they will be used on a per-circuit basis. If none is provided, the
```
I guess I have a similar question around these sentences like I had above. And here I see fidelity's but am missing seeing such an object

> If none is provided, the
> fidelity's default number of shots will be used for all circuits. If this number is
>  also set to None, the underlying primitive's default number of shots will be used
> for all circuits.

tnemoz replied:
I'll update the docstring when I have the confirmation on the other comment that my proposal for the docstring works for you!

woodsp-ibm commented on qiskit_algorithms/minimum_eigensolvers/qaoa.py:
```diff
@@ -117,13 +125,19 @@ def __init__(
         self.initial_state = initial_state
         self._cost_operator = None
+        self._transpiler = transpiler
+        self._transpiler_options = transpiler_options if transpiler_options is not None else {}
```
Ah I see here that there is special treatment for None where by we create an empty dict that ** will be ok on (see an earlier item comment around this in the ampitude estimators)

tnemoz replied:
Yes, I try to avoid defining an empty `dict` as a default value to avoid any nasty side-effect!

woodsp-ibm reviewed (commented):
I quickly skimmed over the code for some initial observations - that's a lot of changes, thanks for persisting with this

tnemoz commented:
Also, the tests are failing because of the issues that are mentioned in the table in the first message. I thought it was OK because it only concerned a version of Qiskit that wasn't released yet, but 2.1.0 was released hours ago and the issues that I've mentioned are not fixed in this version...

What is the best course of action here?

If it helps, the tests for Qiskit 1.4 and 2.0 should pass.

woodsp-ibm commented:
For the notebook timeout its configured in conf.py here https://github.com/qiskit-community/qiskit-algorithms/blob/bc071ca43883b2b8d2232b0574322d7a7cdeca2e/docs/conf.py#L89 so you should be able to increase that is that's what it needs.

So CI using 2.1 is now causing tests to fail because of bugs it has that are not present in 2.0.x. Maybe in `requirements-dev.txt` a line disallowing Qiskit 2.1.0 could be added for now, I think that should work in conjunction with what's in requirements - of course if 2.1.1 does not fix things... maybe it needs to cover != 2.1.* to be safe (assumes 2.2 works of course!) or just restrict to < 2.1 for now - bit of a pain!

tnemoz commented:
> Maybe in `requirements-dev.txt` a line disallowing Qiskit 2.1.0 could be added for now

I've put a commit doing this, but may I ask why should we change it in requirements-dev rather than requirements? People using qiskit-algorithms with Qiskit 2.1 may be exposed to the same bug don't they? Or is there something I haven't understood?

Thanks for the quick answers in all cases!

woodsp-ibm commented:
FYI #233 which may well impact any future runs of this too.

tnemoz commented:
> FYI #233 which may well impact any future runs of this too.

Is it OK to force `scipy<1.16.0` in `requirements.txt` for now and fix this issue later on?

woodsp-ibm commented:
CI and the notebook now pass under I fixed the failure. However as you had changed a couple of the files it seems that the changes overlapped and cause a conflict. 

For the notebook I just edited and saved it and did not save the re-run version so that I did not get depercation warnings in what was checked in. It did update the Python version when doing that which conflicts but having it be the version you ran things under is fine.

For the unit test I guess you had changed the initial point - hopefully the one I had changed it too works fine now instead. But I;ll let you resolve the conflicts here

woodsp-ibm commented:
It seems with the newer primitives and the values I had found that worked across scipy 1.15.3 and 1.16.0 with the former primitives (that you can see work fine in the nightly CI actions here), that a couple of tests did not do so well with 1.16.0 here. Hopefully you can find values that pass both with 1.15.3 and 1.16.0 the newer primitives here. For me I ran the test locally uninstalling/installing scipy to get the version I wanted and trial and error finding something that worked with both - on the positive side its only a couple of tests failing so some of the changes I had to make in other tests worked out ok it seems.

tnemoz commented:
@woodsp-ibm 
There is still a VQD test fail in the `test_callback` function. However, since the point of this test is to check that the callback works as intended, wouldn't it be simpler to not use COBYLA for testing, and use another optimizer like `SLSQP` or `SPSA`?

woodsp-ibm commented:
> wouldn't it be simpler to not use COBYLA for testing, and use another optimizer like SLSQP or SPSA?

My recollection is that I tried with SLSQP and given it computes a finite diff gradient using the objective function I think I ended up with so many callbacks and the array was large. Maybe since SPSA only does + and - around the point I guess you could try that. This test as you can see I could not get to work with the same values for each version rather I ended up detecting the scipy version and to get something that would work. If SPSA seems to do it that might be better as it's not from scipy so should not dependent code like I ended up with in the test.

tnemoz commented:
Sicne the only points that diverged were the last two ones and since the goal of this test is to check that the callback works as expected, I reduced the number of iterations of `COBYLA` from 12 to 10 and now it seems to pass with both `scipy` versions. Is this solution acceptable?

woodsp-ibm commented:
10 is fine - it was less (3) before but its seems that the newer COBYLA does not stop on so few iters despite it seemingly being a limit ie maxiter - maybe that will change but this seems good.

tnemoz commented:
And another weird thing: the CLA complained that some commits done by Léna weren't associated to an email address that signed the CLA when the PR was still a draft, but doesn't seem to complain now that we've put it ready for review, is it expected?

woodsp-ibm commented:
> And another weird thing: the CLA complained that some commits done by Léna weren't associated to an email address that signed the CLA when the PR was still a draft, but doesn't seem to complain now that we've put it ready for review, is it expected?

I think the CLA is for qiskit projects overall so could they have signed it on another one. If CLA passes it all seems good to me.

tnemoz commented:
@woodsp-ibm We've added the release note, let us know if that's not what you expected!

woodsp-ibm commented:
I am thinking that perhaps adding a link to this migration guide in the release notes might be helpful https://quantum.cloud.ibm.com/docs/en/migration-guides/v2-primitives It not only covers the runtime ones but also has a link to the v1 V2 migration for reference primitives in Qiskit. 

Do you think an Estimator sample of before with V1 and now with V2 would be helpful as well, alongside the Grover Sampler one. That would cover both primitive types and their usage in algorithms from a migration perspective.

One minor text item

> Replace all instances of V1 primitives by their V2 counterpart.

Would this be better i.e. `usage` versus `instances`

> Replace all usage of V1 primitives by their V2 counterpart.


Overall though it seems fine to me.

tnemoz commented:
We've updated the release note and also removed a section in another release note stating that `qiskit-algorithms` wasn't supporting Qiskit 2.0, since this is now the case.

woodsp-ibm commented on releasenotes/notes/replaced-v1-primitives-by-v2-ones-897d9610351bd54b.yaml:
```diff
@@ -0,0 +1,109 @@
+    :class:`~qiskit.circuit.QuantumCircuit` or a list thereof and
+    additional options and returns the transpiled version of its input(s)
+    according to the provided options.
+features:
```
I would agree that the latter option seems to be the one for now - perhaps there is room for optimization at some point in the future if the more verbose way seems to need it. If each item can be used standalone, if it needs to transpile circuits as it going to do execution itself then I agree the easiest option is to have the transpilation on each item. This would also allow one to do different transpiles if that has any value which trying to pass options on down would not.

When you talk about AdaptVQE does it need any transpilation option itself or is that all done by VQE as its the one doing the execution. I do see AdaptVQE is direct setting the VQE ansatz public attribute, which I guess anyone can. I imagine if it changed that would need to trigger a re-transpilation assuming the transpiled circuit is cached.

tnemoz replied:
While adapting the code to include this change, I noticed that `aux_operators` is treated differently in `VQE` and in `VQD`. In `VQE`, [one simply has](https://github.com/qiskit-community/qiskit-algorithms/blob/bd91342d09b0b8682d7cc2993035a6f45fdac47c/qiskit_algorithms/minimum_eigensolvers/vqe.py#L213C9-L221C43)

```python
if aux_operators is not None:
    aux_operators_evaluated = estimate_observables(
        self.estimator,
        self.ansatz,
        aux_operators,
        optimizer_result.x,  # type: ignore[arg-type]
    )
else:
    aux_operators_evaluated = None
```

wihle in `VQD` [more checks are performed](https://github.com/qiskit-community/qiskit-algorithms/blob/bd91342d09b0b8682d7cc2993035a6f45fdac47c/qiskit_algorithms/eigensolvers/vqd.py#L215C9-L234C33)

```python
if aux_operators:
    zero_op = SparsePauliOp.from_list([("I" * self.ansatz.num_qubits, 0)])

    # Convert the None and zero values when aux_operators is a list.
    # Drop None and convert zero values when aux_operators is a dict.
    key_op_iterator: Iterable[tuple[str | int, BaseOperator]]
    if isinstance(aux_operators, list):
        key_op_iterator = enumerate(aux_operators)
        converted: ListOrDict[BaseOperator] = [zero_op] * len(aux_operators)
    else:
        key_op_iterator = aux_operators.items()
        converted = {}
    for key, op in key_op_iterator:
        if op is not None:
            converted[key] = zero_op if op == 0 else op  # type: ignore[index]

    aux_operators = converted

else:
    aux_operators = None
```

What's the reason for that? In particular, since I need to iterate over `aux_operators` to apply the ansatz layout to the operators, I end up writing a very similar code for VQE, including defining  a `zero_op` placeholder for instance (see the last commit). Is that expected? Since I don't know what's the purpose of these `aux_operators`, I'm a bit lost as to what should I do with them.

woodsp-ibm replied:
The aux_operators were often used in Nature where for instance we would be computing the ground state of the molecule using the main electronic operator but the molecule has other properties that can be measured in that ground state - where some for instance the number of particles operator can be used as check for that state since that would be a known quantity. I recall operators being X,Y and Z dipole operators, the number of particles operator and magnetization. I seem to recall 6 but hopefully that gives the idea - extra operators we just compute the expectation values of given the state found.

ListOrDict type says the type in the List is Optional so None is allowed. My recollection was that the aux operators, which the user could supply, in the process of computation we may, for example in Nature find symmetries of the molecule which using a symmetry operation we could then reduce the size of the main operator without losing an precision (i.e. remove the symmetries from it to make it smaller). However while this symmetry may hold for the main operator it might not for an aux_operator, So if it did not commute with symmetry we would have to remove it as the size of the main operator, being reduced by symmetries, would no longer match so we could not use the resultant state to estimate that aux_operator expectation value. In the list they were positional so to maintain the positions down and back up the stack when processing, so as things ended up back to the user as expected, Nones were allowed in there. Part of the reason a Dict form was added later that the key/value pairs were easier to drop and still understand what the dict had in it.

Why one handles it better than the other I am not sure. Things have been changed quite a bit in algorithms as things evolved. Does that help at all?

woodsp-ibm commented:
>  It seems like Sphinx complains about a duplicate description for the ansatz attribute of VQE, VQD and PVQD but we're not too sure about what we should do about it. Do you have any guidance on this?

`ansatz` was a public attribute in those classes right, and you made it private and added an explicit getter and setter which both have docstrings on them. Before - and its still there - such public attributes were described in the ` Attributes:` section that precedes the init method. Given ansatz is now done via the documented getter/setter I think deleting the entry from the list in `Attributes:` should fix things.

tnemoz commented:
> It seems its still failing the same with ansatz as it looks like you removed `anstaz` from the `Args` section rather than the `Attributes` section. It should be in `Args`, as its a parameter to the init constructor, just not in `Attributes` any more. The Attributes section is just above the init methid describing public attribues of the class. E.g. line 94 in your vqd.py

Yeah, not my brightest moment :')

tnemoz commented:
@woodsp-ibm By the way, there's a test in in `test_sampler_gradient.py` that we skipped because of the time it was taking to run. However, we're not too sure about how relevant this test is in the first place. Should we uncomment the skip, which would quite drastically increase the tests running time, or should we remove the test altogether?

a-matsuo commented:
@woodsp-ibm Hey Steve, it’s been a while. I can’t figure out why the tests are taking so long to run. Since ParamShiftSamplerGradient and LinCombSamplerGradient need to reconstruct multiple circuits, maybe that’s where the slowdown is happening. Sorry I can’t be more helpful.

tnemoz commented:
> @tnemoz there has been no input to the above question around the unit test, I believe all the work that was intended for this PR is now done right so otherwise its ready to go right?

Hey @woodsp-ibm, sorry for the time without news!

I think it's up to you! As things currently are, `AdaptVQE` doesn't support transpiling, as transpiling a circuit would later fail the test that the ansatz is of the `EvolvedOperatorAnsatz` type (as per the discussion in #231).

We also intended to fix the issues in the Monitoring Qiskit workflow, but there are still some issues that are not fixed, even though it seems PR are on their way.

All in all, it probably depends on whether you intend to release a new version of `qiskit-algorithms` right after this PR. If so, then maybe we could add a warning in the release notes that `AdaptVQE` doesn't support being passed a transpiler for now (which effectively prevents using it on a real backend when being passed an `aux_operators`). If not, then this could be done at the same time when #231 is fixed.

Concerning the monitoring, AFAIK this PR will work with Qiskit up to version 2.1, and maybe an issue could be opened to tackle this?

woodsp-ibm commented:
> We also intended to fix the issues in the Monitoring Qiskit workflow, but there are still some issues that are not fixed, even though it seems PR are on their way.

The Monitoring task was done to just give us a heads up on what was happening as far as Qiskit was concerned by installing/testing against their main branch. As its work in progress there can be issues that arise that cause things to fail here - and we have given feedback in the past when things were more active here in case the issue we saw was unknown.

> Concerning the monitoring, AFAIK this PR will work with Qiskit up to version 2.1, and maybe an issue could be opened to tackle this?

In looking at `requirements.txt` its set to qiskit >=1.0 but we limit the testing in `requirements-dev.txt` to not include 2.1.* so it does not test on the latest actual release. My recollection was that there were some issues with 2.1.0 but since then 2.1.1 has been released which I do not know if it fixed the issues we had here. If so the restriction could be changed to just 2.1.0 rather than the wild card at end preventing 2.1.1 from being used, Of maybe by `up to version 2.1` you meant not including that though the main requirements.txt, and an end user installing qiskit, it would pick the latest by default, i.e 2.1.1

> All in all, it probably depends on whether you intend to release a new version of qiskit-algorithms right after this PR.

Things can be coordinated around a release -  if its felt more needs doing, can be done, before a release then that's fine. This PR has been going for some time I am sure a bit longer is ok.

And in terms of some issues, like changing to circuit library functions, as their removal from deprecation seems to have been moved out to Qiskit 3.0 it appears there is more time to look at things like that.

woodsp-ibm commented:
> And then we could probably merge this PR with the monitoring tests (and thus, hopefully by then, the next Qiskit release), passing. Would that work for you?

Ah you had raised more specific issues. 14714 was merged last night and in the issue I raised I was asked to retry. This is how I responded

> I just re-ran the two CI jobs (Monitor Qiskit Main in 3.9 and 3.13) that test against Qiskit main branch, in the above referenced PR, and the failures are unchanged and still there.

Do you think the tests should be passing - in looking at the operators we create n say the VQD std dev test I thought they were ok. Now I believe they get converted to some other type in the estimator - since you have the code locally does it look like the observables passed to the estimator at least are ok?

But yes you might like to add the pin back again to requirements-dev - we can remove it once things seem to pass for main and a newer version of qiskit has been released that fixes things (assuming things are all ok on the algs side of course)

tnemoz commented:
So, by modifying `evaluate_observables` to bypass the Estimator altogether when the input observable is nil, it seems to solve most of the failing tests in the monitoring task. Out of the remaining four, one of them is solved by the changes implemented in https://github.com/Qiskit/qiskit/pull/14611.

If I'm not mistaken, the remaining three are due to the fact that we directly use the `run` method of an `Estimator` instead of passing it to `evaluate_observables`. I understand that this function was initially destined for another usage, but I think it could be a general tool for evaluating expectation values.

This may however pose some problems. For instance, in the `_run_unique` method of `LinCombEstimatorGradient`, one would have to replace
```python
        job = self._estimator.run(pubs)

        try:
            results = job.result()
        except AlgorithmError as exc:
            raise AlgorithmError("Estimator job failed.") from exc
```
by
```python
        results = [
            estimate_observables(
                self._estimator,
                quantum_state,
                [observable],
                parameters_values,
                threshold=-float("inf")
            )[0][0]
            for (quantum_state, observable, parameters_values) in pubs
        ]
```
However, a direct consequence of doing it like this is that it not only launches multiple jobs, but also waits for each job to end before launching the next one, which of course results in a large overhead.

The solution would be to modify `estimate_observables` to accept PUB-like inputs, but I'm not too sure what how that would go, especially with the `ListOrDict` typing. Another solution which would be much easier to work with would be to create a second function, say `run_estimator_job` that would be a glorified `run` method of an `Estimator`, which would additionally bypass nil observables. This would make it much easier to make it accept PUB-like inputs I think. What are your views on this?

woodsp-ibm commented:
> Another solution which would be much easier to work with would be to create a second 

A different method like that seems more promising I agree. With the usage being inside algorithms calling this and result, instead of directly on an estimator seems reasonable enough, if that's what it takes,

tnemoz commented:
> @tnemoz Thanks so much for doing this. It was a lot of work no doubt and thanks too for raising and sorting out the issues here with the latest Qiskit release. The monitor main should once again pass, as you say, when the last of those issues around the complex coeffs is fixed in 2.1.2. Tremendous piece of work, thanks once again.

Thanks! It was fun (mostly, I now hate the Sampler gradients) working on this 😄 And thank you for the reactivity on answering my questions!

I'll probably try to continue contributing on other issues now that I have a good vision of the codebase :)

By the way, I don't think I have the rights to merge the PR by myself, even now that you've approved it!

tnemoz commented:
Well, this PR did take a year to complete, but that's a funny bug :smile:
